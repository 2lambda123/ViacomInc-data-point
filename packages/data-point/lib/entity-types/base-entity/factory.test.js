/* eslint-env jest */

const Factory = require('./factory')

describe('validateResolve', () => {
  const entity = {
    id: 'foo:bar'
  }
  it('should throw error if type is not function', () => {
    expect(() => {
      Factory.validateResolve(entity, 'string')
    }).toThrowErrorMatchingSnapshot()
  })
  it('should throw error if function does not match arity', () => {
    expect(() => {
      Factory.validateResolve(entity, a => {})
    }).toThrowErrorMatchingSnapshot()
  })
  it('should resolve resolve method if no errors found', () => {
    const resolver = (a, b) => {}
    expect(Factory.validateResolve(entity, resolver)).toEqual(resolver)
  })
})

describe('Factory.createEntityType', () => {
  test('It should create entity defaults', () => {
    const resolve = (a, b) => true
    const entityType = {
      resolve,
      before: '$',
      value: '$',
      error: '$',
      after: '$'
    }
    const entity = Factory.createEntityType('foo', 'bar', entityType)

    expect(entity).toHaveProperty('isEntityInstance', true)
    expect(entity).toHaveProperty('id', 'foo:bar')
    expect(entity).toHaveProperty('name', 'bar')
    expect(entity).toHaveProperty('resolve', resolve)
    expect(entity).toHaveProperty('value.type', 'ReducerPath')
    expect(entity).toHaveProperty('error.type', 'ReducerPath')
    expect(entity).toHaveProperty('after.type', 'ReducerPath')
    expect(entity).toHaveProperty('params', {})
  })
})

describe('EntityFactory', () => {
  const resolve = (a, b) => true
  const create = () => {
    return { resolve }
  }
  it('should create Entity Factory', () => {
    const BarFactory = Factory.create('bar', create)
    expect(BarFactory).toBeInstanceOf(Function)
    expect(BarFactory.length).toEqual(2)
  })
  it('should create new entity instance', () => {
    const BarFactory = Factory.create('bar', create)
    const entity = BarFactory('foo', {})
    expect(entity).toMatchSnapshot()
  })
  it('should create new entity instance with generic name', () => {
    const BarFactory = Factory.create('bar', create)
    const entity = BarFactory({})
    expect(entity).toMatchSnapshot()
  })
})
