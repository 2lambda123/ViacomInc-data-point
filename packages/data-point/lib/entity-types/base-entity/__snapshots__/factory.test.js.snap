// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityFactory should create new entity instance 1`] = `
Object {
  "entityType": "bar",
  "id": "bar:foo",
  "isEntityInstance": true,
  "name": "foo",
  "params": Object {},
  "resolve": [Function],
}
`;

exports[`EntityFactory should create new entity instance with generic name 1`] = `
Object {
  "entityType": "bar",
  "id": "bar:generic",
  "isEntityInstance": true,
  "name": "generic",
  "params": Object {},
  "resolve": [Function],
}
`;

exports[`validateResolve should throw error if function does not match arity 1`] = `"Entity \\"foo:bar\\" resolve(accumulator:Accumulator, resolveReducer:Function) function must have an arity of 2"`;

exports[`validateResolve should throw error if type is not function 1`] = `"Entity \\"foo:bar\\" must be provided with a resolve(accumulator:Accumulator, resolveReducer:Function) function"`;
